include/lsst/afw/math/statistics/Framework.h

```diff
@@ -0,0 +1,615 @@
+          typename WeightPixelT>
+void accumulateHelper(ProductSet<Args...>& t, ImagePixelT img, MaskPixelT msk, VariancePixelT var,
+                      WeightPixelT wgt) {
+    AccumulateHelper<decltype(t), sizeof...(Args)>::accumulate(t, img, msk, var, wgt);
+}
+
+/**
+ * Helper class to call `accumulateClipped` member function for every member of a `ProductSet`.
+ */
+template <typename T, size_t N>
+struct AccumulateClippedHelper {
+    template <typename ImagePixelT, typename MaskPixelT, typename VariancePixelT, typename WeightPixelT>
+    static void accumulateClipped(T& t, ImagePixelT img, MaskPixelT msk, VariancePixelT var,
+                                  WeightPixelT wgt) {
+        AccumulateClippedHelper<T, N - 1>::accumulateClipped(t, img, msk, var, wgt);
+        std::get<N - 1>(t).accumulateClipped(img, msk, var, wgt);
+    }
+};
+
+template <typename T>
+struct AccumulateClippedHelper<T, 1> {
+    template <typename ImagePixelT, typename MaskPixelT, typename VariancePixelT, typename WeightPixelT>
+    static void accumulateClipped(T& t, ImagePixelT img, MaskPixelT msk, VariancePixelT var,
+                                  WeightPixelT wgt) {
+        std::get<0>(t).accumulateClipped(img, msk, var, wgt);
+    }
+};
+
+/**
+ * Call `accumulateClipped` member function for every member of `ProductSet` argument `t`.
+ */
+template <typename... Args, typename ImagePixelT, typename MaskPixelT, typename VariancePixelT,
+          typename WeightPixelT>
+void accumulateClippedHelper(ProductSet<Args...>& t, ImagePixelT img, MaskPixelT msk, VariancePixelT var,
+                             WeightPixelT wgt) {
+    AccumulateClippedHelper<decltype(t), sizeof...(Args)>::accumulateClipped(t, img, msk, var, wgt);
+}
+
+/**
+ * Helper class to call `combine` member function for every member of a `ProductSet`.
+ */
+template <typename T, size_t N>
+struct CombineHelper {
+    template <typename Product>
+    static void combine(T& t, Product const& other) {
+        CombineHelper<T, N - 1>::combine(t, other);
+        std::get<N - 1>(t).combine(other);
+    }
+};
+
+template <typename T>
+struct CombineHelper<T, 1> {
+    template <typename Product>
+    static void combine(T& t, Product const& other) {
+        std::get<0>(t).combine(other);
+    }
+};
+
+/**
+ * Call `combine` member function for every member of `ProductSet` argument `t`.
+ */
+template <typename... Args, typename Product>
+void combineHelper(ProductSet<Args...>& t, Product const& other) {
+    CombineHelper<decltype(t), sizeof...(Args)>::combine(t, other);
+}
+
+/**
+ * Base class for statistics products (e.g. Sum, Count, Variance, etc.).
+ *
+ * A Product is templated on the intermediate Products it requires (or None),
+ * and defines an `accumulate` member function (default noop).
+ */
+template <typename... Ts>
+class Product {
+public:
+    using Intermediate = typename FlattenOne<ProductSet<Ts...>>::type;
```

Comment:
I'm surprised that you flatten only one level. Is this easy to explain? Does this fully avoid computing the same statistic multiple times for different products?